text/0096-contract-metadata.md

```diff
@@ -0,0 +1,206 @@
+        let uref = storage::new_uref(description);
+        runtime::put_key(DESCRIPTION_KEY, uref.into());
+    }
+
+    if let Some(icon_uri) = icon_uri {
+        let uref = storage::new_uref(icon_uri);
+        runtime::put_key(ICON_URI_KEY, uref.into());
+    }
+
+    if let Some(project_uri) = project_uri {
+        let uref = storage::new_uref(project_uri);
+        runtime::put_key(PROJECT_URI_KEY, uref.into());
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn contract_name() {
+    let value: Option<String> = read_metadata(NAME_KEY);
+    runtime::ret(CLValue::from_t(value).unwrap_or_revert());
+}
+
+#[no_mangle]
+pub extern "C" fn contract_description() {
+    let value: Option<String> = read_metadata(DESCRIPTION_KEY);
+    runtime::ret(CLValue::from_t(value).unwrap_or_revert());
+}
+
+#[no_mangle]
+pub extern "C" fn contract_icon_uri() {
+    let value: Option<String> = read_metadata(ICON_URI_KEY);
+    runtime::ret(CLValue::from_t(value).unwrap_or_revert());
+}
+
+#[no_mangle]
+pub extern "C" fn contract_project_uri() {
+    let value: Option<String> = read_metadata(PROJECT_URI_KEY);
+    runtime::ret(CLValue::from_t(value).unwrap_or_revert());
+}
+
+// Helper function for reading metadata from named URefs
+fn read_metadata(key_name: &str) -> Option<String> {
+    runtime::get_key(key_name)
+        .and_then(|key| key.into_uref())
+        .and_then(|uref| storage::read::<String>(uref).unwrap_or(None))
+}
+```
+
+**Explanation**:
+- The `init` function creates and stores metadata URefs only once at contract deployment.
+- Metadata retrieval functions (`name`, `description`, etc.) expose the stored data.
+
+**Example Usage**: If a user (or another contract) calls the `name()` entry point (via an RPC query or contract call), the contract will respond with either `Some("Casper DEX")` (for example) or `None` if no name was set. Similarly, calling `icon_uri()` might return `Some("https://example.com/casperdex.png")` which a UI can use to load the icon, or `None` if the contract didn’t specify an icon.
+
+Developers can integrate this trait easily. For instance, a contract written using the Odra framework or similar could derive or implement the `CEP96ContractMetadata` trait, and internally it could manage the storage as shown. The important part is that the compiled contract exports the four functions with the correct signatures.
+
+## Recommended Client Usage (Wallets, Explorers, Indexers)
+
+[recommended-client-usage-wallets-explorers-indexers]: #recommended-client-usage-wallets-explorers-indexers
+
+To adopt this standard, tools and services in the Casper ecosystem should:
+
+- Detect compliance by checking for named keys (`contract_name`, `contract_description`, `contract_icon_uri`, `contract_project_uri`) in the contract's state via Casper RPC (`query_global_state`).
+- Directly query metadata via the Casper RPC interface without performing contract calls, leveraging the efficient state queries.
```

Comment:
You could merge those two points as 2nd point is already covered in 1st point (`query_global_state` is the key)